### PR TITLE
[WIP] Boost by downloads

### PR DESCRIPTION
**Pull request by yishaigalatzer**
This is a WIP suggestion for boosting by download counts.

I think we should have a more efficient way to get to the download count, but I didn't want to make it a big change at the moment. This is more for discussion with @maartenba 

Things to figure out
1. More test cases
2. Running against production data
3. Tweaking the values
4. Seeing if we can deprecate the ranking and use download counts instead

### CODE REVIEW

File: src/NuGet.Indexing/Downloads.cs [hunk 1]
```diff
@@ -2,18 +2,37 @@
...
 using System.Linq;
 using Microsoft.Extensions.Logging;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
 using FrameworkLogger = Microsoft.Extensions.Logging.ILogger;
 
 namespace NuGet.Indexing
 {
-    public static class Downloads
+    public class Downloads
```
Comment (yishaigalatzer) [inline src/NuGet.Indexing/Downloads.cs]: the changes here are mostly to keep the code readable and sane. I moved away from IDictionary to a concrete class with just the needed method

File: src/NuGet.Indexing/DownloadsByVersion.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace NuGet.Indexing
+{
+    public class DownloadsByVersion
```
Comment (yishaigalatzer) [inline src/NuGet.Indexing/DownloadsByVersion.cs]: sanness

File: src/NuGet.Indexing/GalleryServiceImpl.cs [hunk 3]
```diff
@@ -68,9 +68,15 @@ private static void DocumentCountImpl(JsonWriter jsonWriter, IndexSearcher searc
...
+        private static void ListDocumentsImpl(JsonWriter jsonWriter,
+            NuGetIndexSearcher searcher,
+            Query query,
+            string sortBy,
+            int skip,
+            int take,
+            NuGetSearcherManager manager)
         {
-            Query boostedQuery = new RankingScoreQuery(query, searcher.Rankings);
+            Query boostedQuery = new RankingScoreQuery(query, searcher.Downloads, searcher.Rankings);
```
Comment (yishaigalatzer) [inline src/NuGet.Indexing/GalleryServiceImpl.cs]: This is threading downloads down to the booster/scorer

File: src/NuGet.Indexing/OwnersResult.cs [hunk 4]
```diff
@@ -0,0 +1,24 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+
+namespace NuGet.Indexing
```
Comment (yishaigalatzer) [inline src/NuGet.Indexing/OwnersResult.cs]: got rid of nested classes. Public classes should not be nested. And one file per class

File: src/NuGet.Indexing/Properties/Strings.resx [hunk 5]
```diff
@@ -141,6 +141,14 @@
     "universal windows platform"
   ],
 
+  "json.net" : [
```
Comment (yishaigalatzer) [inline src/NuGet.Indexing/Properties/Strings.resx]: tried more acronyms, but that doesn't seem to actually get every called.

File: src/NuGet.Indexing/Properties/Strings.resx [hunk 6]
```diff
@@ -158,4 +166,7 @@
   ]
 }</value>
   </data>
+  <data name="String1" xml:space="preserve">
```
Comment (yishaigalatzer) [inline src/NuGet.Indexing/Properties/Strings.resx]: I'll remove

File: src/NuGet.Indexing/Ranking.cs [hunk 7]
```diff
@@ -0,0 +1,11 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace NuGet.Indexing
```
Comment (yishaigalatzer) [inline src/NuGet.Indexing/Ranking.cs]: nested class

File: src/NuGet.Indexing/RankingsHandler.cs [hunk 8]
```diff
@@ -12,69 +12,66 @@ namespace NuGet.Indexing
...
     {
         private readonly IDictionary<string, int> _rankings;
+        private RankingBySegmentReader _rankingBySegmentReaderName;
 
-        private IDictionary<string, Ranking[]> _rankingTuples;
         public RankingResult Result { get; private set; }
 
         public RankingsHandler(IDictionary<string, int> rankings)
         {
+            if (rankings == null)
```
Comment (yishaigalatzer) [inline src/NuGet.Indexing/RankingsHandler.cs]: Added a check, this was actually surprisingly throwing ...

File: src/NuGet.Indexing/RankingsScoreQuery.cs [hunk 9]
```diff
@@ -13,85 +13,113 @@ public class RankingScoreQuery : CustomScoreQuery
...
                 // document higher in search results.
-                float score = GetRankingScore(_readerName, _rankings, doc, _baseBoost);
-                if (score == 0.0f)
-                {
-                    return subQueryScore;
-                }
-
-                float adjustedScore = subQueryScore * score;
-                return adjustedScore * 10;
+                var adjustedScore = (double)subQueryScore;
```
Comment (yishaigalatzer) [inline src/NuGet.Indexing/RankingsScoreQuery.cs]: core change

File: src/NuGet.Services.BasicSearch.Cloud/NuGet.Services.BasicSearch.Cloud.ccproj [hunk 10]
```diff
@@ -4,7 +4,7 @@
   <PropertyGroup>
     <Configuration Condition=" '$(Configuration)' == '' ">Debug</Configuration>
     <Platform Condition=" '$(Platform)' == '' ">AnyCPU</Platform>
-    <ProductVersion>2.7</ProductVersion>
+    <ProductVersion>2.9</ProductVersion>
```
Comment (yishaigalatzer) [inline src/NuGet.Services.BasicSearch.Cloud/NuGet.Services.BasicSearch.Cloud.ccproj]: I can roll this back, or we can move to SDK 2.9 properly. BTW 2.7 is no longer available to install by default

File: src/NuGet.Indexing/Downloads.cs (hunk 1 above)
Comment (maartenba) [inline src/NuGet.Indexing/Downloads.cs]: :+1: me likes

File: src/NuGet.Indexing/GalleryServiceImpl.cs [hunk 11]
```diff
@@ -68,9 +68,15 @@ private static void DocumentCountImpl(JsonWriter jsonWriter, IndexSearcher searc
...
+        private static void ListDocumentsImpl(JsonWriter jsonWriter,
+            NuGetIndexSearcher searcher,
+            Query query,
+            string sortBy,
+            int skip,
+            int take,
+            NuGetSearcherManager manager)
         {
-            Query boostedQuery = new RankingScoreQuery(query, searcher.Rankings);
+            Query boostedQuery = new DownloadsBoostedQuery(query, searcher.DocIdMapping, searcher.Downloads, searcher.Rankings);
```
Comment (maartenba) [inline src/NuGet.Indexing/GalleryServiceImpl.cs]: Think we can get rid of the Rankings code path/parameter? Or is it still needed?

File: src/NuGet.Indexing/Handlers/SegmentToMainReaderMappingHandler.cs [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+using System.Collections.Generic;
+using Lucene.Net.Documents;
+using Lucene.Net.Index;
+using NuGet.Versioning;
+
+namespace NuGet.Indexing
+{
+    public class SegmentToMainReaderMappingHandler : IIndexReaderProcessorHandler
+    {
+        private readonly Dictionary<string, int[]> _mapping = new Dictionary<string, int[]>(StringComparer.Ordinal);
```
Comment (maartenba) [inline src/NuGet.Indexing/Handlers/SegmentToMainReaderMappingHandler.cs]: Do you see a possibility to somehow keep the dictionary around? Per index reload we're recreating a few 100k's of variables, which is bound to induce a GC.

File: src/NuGet.Indexing/Properties/Strings.resx (hunk 5 above)
Comment (maartenba) [inline src/NuGet.Indexing/Properties/Strings.resx]: The `.` is the culprit - the `DotTokenizer` is first in the chain, which means there are never terms with a `.` coming through here.

File: src/NuGet.Indexing/Querys/DownloadsScoreProvider.cs [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+            // Package might not exist in the package table hence the ?.Total
+            var totalDownloads = _downloads[indexId]?.Total ?? 0;
+
+            float downloadAdjuster = DownloadScore(totalDownloads);
+
+            return downloadAdjuster;
+        }
+
+        public static float DownloadScore(long totalDownloads,
+            int factor = 5, // 20
```
Comment (maartenba) [inline src/NuGet.Indexing/Querys/DownloadsScoreProvider.cs]: Is this going to change often?

File: src/NuGet.Services.BasicSearch.Cloud/NuGet.Services.BasicSearch.Cloud.ccproj (hunk 10 above)
Comment (maartenba) [inline src/NuGet.Services.BasicSearch.Cloud/NuGet.Services.BasicSearch.Cloud.ccproj]: We can do the upgrade (but that requires upgrading all build servers as well). I wish the SDK was a NuGet package as all we need is one DLL... Your call :-)

File: src/NuGet.Services.BasicSearch/Web.config [hunk 14]
```diff
@@ -14,11 +14,9 @@
     <sessionState mode="Off" />
   </system.web>
   <appSettings>
-    <!--
-    <add key="Local.Lucene.Directory" value="{path to local Lucene directory}" />
-    <add key="Local.Data.Directory" value="{path to local data directory}" />
-    -->
-    
+    <add key="Local.Lucene.Directory" value="C:\nuget\SearchData\Lucene-v2v3" />
```
Comment (maartenba) [inline src/NuGet.Services.BasicSearch/Web.config]: Please revert this one when this PR gets merged

**Comment by maartenba** [comment]
Few comments, really like the changes!

**Comment by maartenba** [comment]
Should we add Rx -> Reactive as an acronym?

File: src/NuGet.Indexing/Properties/Strings.resx (hunk 5 above)
Comment (yishaigalatzer) [inline src/NuGet.Indexing/Properties/Strings.resx]: Ah, I'll look at changing the order then

File: src/NuGet.Services.BasicSearch.Cloud/NuGet.Services.BasicSearch.Cloud.ccproj (hunk 10 above)
Comment (yishaigalatzer) [inline src/NuGet.Services.BasicSearch.Cloud/NuGet.Services.BasicSearch.Cloud.ccproj]: I don't mind, I moved to 2.9 for my local build, you can roll it back before publishing. Moving to 2.9 is lower pri at this point

File: src/NuGet.Indexing/Handlers/SegmentToMainReaderMappingHandler.cs (hunk 12 above)
Comment (yishaigalatzer) [inline src/NuGet.Indexing/Handlers/SegmentToMainReaderMappingHandler.cs]: There shouldn't be 100K variables, just one per segment reader. Need to validate of course

File: src/NuGet.Services.BasicSearch.Cloud/NuGet.Services.BasicSearch.Cloud.ccproj (hunk 10 above)
Comment (maartenba) [inline src/NuGet.Services.BasicSearch.Cloud/NuGet.Services.BasicSearch.Cloud.ccproj]: Going to do the 2.9 update in all projects (gallery and this one) on friday and make sure build server supports it all.

**Comment by maartenba** [comment]
Changes look good. Not sure if we can safely drop the rankings?

Have found some other acronyms we may want to add:
* rx -> reactive extensions
* js -> JavaScript

**Comment by maartenba** [comment]
Just verified this also fixes: https://twitter.com/dot_NET_Junkie/status/730059574095515648

**Comment by maartenba** [comment]
Merged. Please create a new branch off of `dev` if you want to WIP on removing rankings.

**Review by qzm6127 — APPROVED** [review]

**Pull request closed without merging**